### PR TITLE
label-based uninstaller

### PR DESCRIPTION
**Pull request by joelddiaz**
the label-based uninstaller tool will take a key=value as the basic filtering/matching mechanism to identify and remove AWS objects matching the tag/value.

there is special handling for untagged items (IAM objects in particular) where before deleting an object the tool will identify and remove any dependent objects that could otherwise not be found with tagging.

the general flow is that goroutines for each top-level-AWS-object type are launched and will spin (with backoff) and retry until each goroutine finds no more objects to delete.

the tool retries in the face of errors calling into the AWS API, but exceptional errors (like internal data manipulation and timeouts) cause the program to exit prematurely.

### CODE REVIEW

**Comment by dgoodwin** [comment]
Regarding the hung process, I used the installer from a few weeks, it might not have the latest if there was something additional that had to be tagged, could that explain? Should we adjust anything to get the output more clear in a hang scenario?

File: contrib/pkg/tfuninstall/tfuninstall.go [hunk 1]
```diff
@@ -0,0 +1,1375 @@
...
+			return nil, fmt.Errorf("AWS credentials secret %v did not contain key %v",
+				secretName, awsCredsSecretAccessKey)
+		}
+
+		awsConfig.Credentials = credentials.NewStaticCredentials(
+			string(accessKeyID), string(secretAccessKey), "")
+	}
+
+	// Otherwise default to relying on the IAM role/ENV vars
+	s, err := session.NewSession(awsConfig)
```
[the remark this answers is not in the document]
Comment (joelddiaz) [reply contrib/pkg/tfuninstall/tfuninstall.go]: i've been running/developing the tool by just setting AWS_PROFILE and running it outside of anything kube-related. the kube/namespace/secret piece was just copied over from other coutil command that use this functionality.

if we don't think we'll use the secrets method to pass in creds, i'll be more than happy to remove these bits

File: contrib/pkg/tfuninstall/tfuninstall.go [hunk 2]
```diff
@@ -0,0 +1,1375 @@
...
+	describeInstancesInput.Filters = append(describeInstancesInput.Filters, &ec2.Filter{
+		Name:   aws.String("instance-state-name"),
+		Values: []*string{aws.String("running")},
+	})
+
+	for {
+		results, err := ec2Client.DescribeInstances(&describeInstancesInput)
+		if err != nil {
+			logger.Debugf("error listing instances: %v", err)
+			return false, nil
```
[the remark this answers is not in the document]
Comment (joelddiaz) [reply contrib/pkg/tfuninstall/tfuninstall.go]: while the most likely possibility whenever we hit any of these 'list AWS object(s)' errors is a permissions issue, i've been treating these 'error while listing' errors as generic retries to also capture the case where we might be getting API throttled

if we 'return false, err', then the error will be considered fatal, and there will be no exponential backoff.

[the remark this answers is not in the document]
Comment (dgoodwin) [reply contrib/pkg/tfuninstall/tfuninstall.go]: Gotcha, thanks!

**Comment by joelddiaz** [comment]
@dgoodwin thanks for the feedback. i've addressed (or at least attempted to) the PR feedback

**Comment by dgoodwin** [comment]
@joelddiaz lets squash, drop WIP, and merge. Thanks!

**Pull request merged**